### PR TITLE
Initial back compat tests

### PR DESCRIPTION
**Pull request by jebrando**

### CODE REVIEW

File: configs/back_compat_functions.1.cmake [hunk 1]
```diff
@@ -0,0 +1,183 @@
+#Copyright (c) Microsoft. All rights reserved.
```
Comment (bikamani) [inline configs/back_compat_functions.1.cmake]: Remove this file?

File: tests/common_test_files/back_compat_test_common.c [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+BACK_COMPAT_HANDLE back_compat_init(TEST_PROTOCOL_TYPE protocol_type)
+{
+    BACK_COMPAT_INFO* result;
+    if (IoTHub_Init() != 0)
+    {
+        LogError("Failure calling IoTHub_Init");
+    }
+    else if ((result = (BACK_COMPAT_INFO*)malloc(sizeof(BACK_COMPAT_INFO))) == NULL)
+    {
+        LogError("Could not initialize IoTHubAccount");
```
Comment (bikamani) [inline tests/common_test_files/back_compat_test_common.c]: Change logging?

File: tests/common_test_files/back_compat_test_common.c [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+        result = NULL;
+    }
+    else
+    {
+        memset(result, 0, sizeof(BACK_COMPAT_INFO));
+        if ((result->tick_cntr_handle = tickcounter_create()) == NULL)
+        {
+            LogError("Could not initialize IoTHubAccount");
```
Comment (bikamani) [inline tests/common_test_files/back_compat_test_common.c]: Change logging?

File: testtools/iothub_test/src/iothub_account.c [hunk 4]
```diff
@@ -0,0 +1,1149 @@
...
+                memset(iothub_account_info->sasDevices, 0, sizeof(IOTHUB_PROVISIONED_DEVICE*) * iothub_account_info->number_of_sas_devices);
+
+#ifdef MBED_BUILD_TIMESTAMP
+                iothub_account_info->connString = getMbedParameter("IOTHUB_CONNECTION_STRING");
+                iothub_account_info->eventhubConnString = getMbedParameter("IOTHUB_EVENTHUB_CONNECTION_STRING");
+                tempCert = getMbedParameter("IOTHUB_E2E_X509_CERT");
+                tempKey = getMbedParameter("IOTHUB_E2E_X509_PRIVATE_KEY");
+                tempThump = getMbedParameter("IOTHUB_E2E_X509_THUMBPRINT");
+#else
+                iothub_account_info->connString = getenv("IOTHUB_CONNECTION_STRING");
```
Comment (bikamani) [inline testtools/iothub_test/src/iothub_account.c:686]: Minor - These two lines can be moved outside of ifdef block

File: testtools/iothub_test/src/iothub_account.c [hunk 5]
```diff
@@ -0,0 +1,1149 @@
...
+                    free(iothub_account_info);
+                    iothub_account_info = NULL;
+                }
+                else if (iothub_account_info->eventhubConnString == NULL)
+                {
+                    LogError("Failure retrieving Event Hub connection string from the environment.\r\n");
+                    free(iothub_account_info);
+                    iothub_account_info = NULL;
+                }
+                /*else if (base64_cert == NULL)
```
Comment (bikamani) [inline testtools/iothub_test/src/iothub_account.c:705]: Do we intend to keep commented checks for future? OR remove it?

File: testtools/iothub_test/src/iothub_account.c [hunk 6]
```diff
@@ -0,0 +1,1149 @@
...
+    {
+        LogError("Failure determining the string length parameters.\r\n");
+        result = MU_FAILURE;
+    }
+    else
+    {
+        if ((accountInfo->iothubName = (char*)malloc(endHost - beginHost + 1)) == NULL)
+        {
+            LogError("Failure allocating iothubName.\r\n");
+            result = MU_FAILURE;
```
Comment (bikamani) [inline testtools/iothub_test/src/iothub_account.c:160]: Probably better to initialize result at line #149 to MU_FAILURE

File: testtools/iothub_test/src/iothub_account.c [hunk 7]
```diff
@@ -0,0 +1,1149 @@
...
+    return result;
+}
+
+static int retrieveConnStringInfo(IOTHUB_ACCOUNT_INFO* accountInfo)
+{
+    int result;
+    int beginName, endName, beginIothub, endIothub, beginHost, endHost, beginKey;
+    size_t totalLen = strlen(accountInfo->connString);
+
+    if (sscanf(accountInfo->connString, "HostName=%n%*[^.]%n.%n%*[^;];%nSharedAccessKeyName=%n%*[^;];%nSharedAccessKey=%n", &beginHost, &endHost, &beginIothub, &endIothub, &beginName, &endName, &beginKey) != 0)
```
Comment (bikamani) [inline testtools/iothub_test/src/iothub_account.c:150]: Add check for null pointer?

File: testtools/iothub_test/src/iothub_account.c [hunk 8]
```diff
@@ -0,0 +1,1149 @@
...
+        }
+        else if ((accountInfo->hostname = (char*)malloc(endIothub - beginHost + 1)) == NULL)
+        {
+            LogError("Failure allocating hostname.\r\n");
+            free(accountInfo->iothubName);
+            result = MU_FAILURE;
+        }
+        else if ((accountInfo->keyName = (char*)malloc(endName - beginName + 1)) == NULL)
+        {
+            LogError("Failure allocating hostName.\r\n");
```
Comment (bikamani) [inline testtools/iothub_test/src/iothub_account.c]: Change logging

File: testtools/iothub_test/src/iothub_account.c [hunk 9]
```diff
@@ -0,0 +1,1149 @@
...
+            free(accountInfo->iothubName);
+            free(accountInfo->hostname);
+            result = MU_FAILURE;
+        }
+        else if ((accountInfo->sharedAccessKey = (char*)malloc(totalLen + 1 - beginKey + 1)) == NULL)
+        {
+            LogError("Failure allocating hostName.\r\n");
```
Comment (bikamani) [inline testtools/iothub_test/src/iothub_account.c]: Change logging

File: testtools/iothub_test/src/iothub_account.c [hunk 10]
```diff
@@ -0,0 +1,1149 @@
...
+                                }
+                                else
+                                {
+                                    if (provisionDevices(iothub_account_info, IOTHUB_ACCOUNT_AUTH_CONNSTRING, iothub_account_info->sasDevices, iothub_account_info->number_of_sas_devices) != 0)
+                                    {
+                                        LogError("Failed to create the SAS device(s)\r\n");
+                                        IoTHubAccount_deinit(iothub_account_info);
+                                        iothub_account_info = NULL;
+                                    }
+                                    //else if (provisionDevice(iothub_account_info, IOTHUB_ACCOUNT_AUTH_X509, &iothub_account_info->x509Device) != 0)
```
Comment (bikamani) [inline testtools/iothub_test/src/iothub_account.c:787]: Should this function have authentication_method argument passed down so less changes for future?

File: testtools/iothub_test/src/iothub_account.c [hunk 11]
```diff
@@ -0,0 +1,1149 @@
...
+
+        free(acctInfo->hostname);
+        free(acctInfo->iothubName);
+        free(acctInfo->iothubSuffix);
+        free(acctInfo->sharedAccessKey);
+        free(acctInfo->sharedAccessToken);
+        free(acctInfo->keyName);
+        free(acctInfo->eventhubAccessKey);
+
+        free(acctInfo->x509Device.deviceId);
```
Comment (bikamani) [inline testtools/iothub_test/src/iothub_account.c:874]: Do we need to free x509Device.x always?

File: testtools/iothub_test/src/iothub_account.c [hunk 12]
```diff
@@ -0,0 +1,1149 @@
...
+                result = NULL;
+            }
+            STRING_delete(cert);
+        }
+        BUFFER_delete(raw_cert);
+    }
+    return result;
+}
+
+IOTHUB_ACCOUNT_INFO_HANDLE IoTHubAccount_Init_With_Config(IOTHUB_ACCOUNT_CONFIG* config, bool testingModules)
```
Comment (bikamani) [inline testtools/iothub_test/src/iothub_account.c:643]: testingModules doesn't seem to be used anywhere. Do we still need it?

File: tests/common_test_files/back_compat_test_common.c [hunk 13]
```diff
@@ -0,0 +1,802 @@
...
+        iterator++;
+    }
+    return result;
+}
+
+static IOTHUB_DEVICE_CLIENT_LL_HANDLE create_device_client(BACK_COMPAT_INFO* back_compat_info, DEVICE_CREATION_TYPE type, IOTHUB_CLIENT_TRANSPORT_PROVIDER protocol)
+{
+    IOTHUB_DEVICE_CLIENT_LL_HANDLE result;
+
+    IOTHUB_PROVISIONED_DEVICE* device_info = IoTHubAccount_GetSASDevice(back_compat_info->acct_info);
```
Comment (bikamani) [inline tests/common_test_files/back_compat_test_common.c:319]: Check for device_info ?

File: CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+set(IOTHUB_REPO_FOLDER ${CMAKE_CURRENT_LIST_DIR}/deps/c-sdk CACHE INTERNAL "Location of the azure c sdk" FORCE)
+include_directories(${IOTHUB_REPO_FOLDER}/iothub_client/inc)
+include_directories(${IOTHUB_REPO_FOLDER}/provisioning_client/inc)
+include_directories(${IOTHUB_REPO_FOLDER}/iothub_service_client/inc)
+
+include_directories(${TESTRUNNERSWITCHER_INC_FOLDER})
+include_directories(${CTEST_INC_FOLDER})
+include_directories(${UMOCK_C_INC_FOLDER})
+
+#????
```
Comment (bikamani) [inline CMakeLists.txt]: Nit: Remove

File: CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+set(CLONE_OPTIONS "-q")
+
+#Clone the test submodules if needed
+if (NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/deps/testrunnerswitcher/CMakeLists.txt")
+    execute_process(COMMAND git clone https://github.com/Azure/azure-c-testrunnerswitcher.git ${CLONE_OPTIONS} ${CMAKE_CURRENT_SOURCE_DIR}/deps/testrunnerswitcher)
+endif()
+if (NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/deps/ctest/CMakeLists.txt")
+    execute_process(COMMAND git clone https://github.com/Azure/azure-ctest.git ${CLONE_OPTIONS} ${CMAKE_CURRENT_SOURCE_DIR}/deps/ctest)
+endif()
+#if (NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/deps/umock-c/CMakeLists.txt")
```
Comment (bikamani) [inline CMakeLists.txt]: Nit: Is this needed?

File: CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+#endif()
+
+if (${refresh_sdk} OR NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/deps/c-sdk/CMakeLists.txt")
+    execute_process(COMMAND git clone https://github.com/Azure/azure-iot-sdk-c.git ${CLONE_OPTIONS} -b ${sdk_branch} --recursive ${CMAKE_CURRENT_SOURCE_DIR}/deps/c-sdk)
+endif()
+
+# Submodules
+add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/deps/testrunnerswitcher)
+
+#add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/deps/umock-c)
```
Comment (bikamani) [inline CMakeLists.txt]: Nit: Remove?

File: testtools/iothub_test/src/iothub_account.c (hunk 6 above)
Comment (jebrando) [reply testtools/iothub_test/src/iothub_account.c:160]: There is a reason I don't, we can talk about it afterwords.

File: testtools/iothub_test/src/iothub_account.c (hunk 7 above)
Comment (jebrando) [reply testtools/iothub_test/src/iothub_account.c:150]: The Null pointer check is on line 692

**Pull request merged**